contracts/paratii/PTIDistributor.sol

```diff
@@ -8,24 +8,45 @@ contract PTIDistributor is Ownable {
 
     Registry public registry;
 
-    event LogDistribute(address _toAddress, uint _amount, string _reason);
+    event LogDistribute(address _toAddress, uint _amount);
+    event LogDebug(bytes32 _hashing);
+    event LogDebugOwner(address _owner);
+
+    mapping(uint256 => bool) public isUsed;
 
     function PTIDistributor(Registry _registry) public {
         owner = msg.sender;
         registry = _registry;
     }
 
-    modifier onlyOwnerOrAvatar() {
-        require((msg.sender == owner) || (msg.sender == registry.getContract('Avatar')));
-        _;
+    function() public payable { }
+
+    function distribute(address _toAddress, uint256 _amount, uint256 _salt, uint8 _v, bytes32 _r, bytes32 _s) {
+      bytes32 message = keccak256(_amount, _salt);
+      bytes memory prefix = "\x19Ethereum Signed Message:\n32";
+      bytes32 prefixedHash = keccak256(prefix, message);
+      require(!isUsed[_salt] && ecrecover(prefixedHash, _v, _r, _s) == owner);
+      isUsed[_salt] = true;
+      ParatiiToken token = ParatiiToken(registry.getContract('ParatiiToken'));
+      token.transfer(msg.sender, _amount);
```

Comment:
use `require` here to make sure the transfer happens or all changes are cancelled.
`require(token.transfer(msg.sender, _amount));`